library/std/src/sys/unix/fs.rs

```diff
@@ -712,22 +701,10 @@ impl Iterator for ReadDir {
                 // requires the full extent of *entry_ptr to be in bounds of the same
                 // allocation, which is not necessarily the case here.
                 //
-                // Absent any other way to obtain a pointer to `(*entry_ptr).d_name`
-                // legally in Rust analogously to how it would be done in C, we instead
-                // need to make our own non-libc allocation that conforms to the weird
-                // imaginary definition of dirent64, and use that for a field offset
-                // computation.
+                // Instead we must access fields individually through their offsets.
                 macro_rules! offset_ptr {
                     ($entry_ptr:expr, $field:ident) => {{
-                        const OFFSET: isize = {
-                            let delusion = MaybeUninit::<dirent64>::uninit();
-                            let entry_ptr = delusion.as_ptr();
-                            unsafe {
-                                ptr::addr_of!((*entry_ptr).$field)
-                                    .cast::<u8>()
-                                    .offset_from(entry_ptr.cast::<u8>())
-                            }
-                        };
+                        const OFFSET: isize = mem::offset_of!(dirent64, $field) as isize;
                         if true {
                             // Cast to the same type determined by the else branch.
                             $entry_ptr.byte_offset(OFFSET).cast::<_>()
```

Comment:
curiosity: I see above it talks about "in bounds of the same allocation, which is not necessarily the case here", but then it uses `byte_offset`, which also has the "inside the same allocation" problem.

Should it perhaps be `wrapping_byte_offset` here?  Or am I misunderstanding the restrictions under which this is operating?

Reply:
I think both the code and comment are correct as written. `byte_offset(OFFSET)` requires the **start** of $field to be in bounds, which it is. `addr_of!((*ptr).$field)` requires the **start and end** of $field to be in bounds which they are not.

Reply:
Ah, thanks!  I hadn't considered the whole width of the field problem.